lib/helm/testclient.go

```diff
@@ -0,0 +1,116 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package helm
+
+import (
+	"github.com/gravitational/gravity/lib/storage"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/trace"
+)
+
+// testClient is the mock Helm client implementation used in tests.
+type testClient struct {
+	// releases keeps track of the installed releases.
+	releases map[string]storage.Release
+	// closed indicates whether the client is closed.
+	closed bool
+	// ClientConfig is the test client configuration.
+	ClientConfig
+}
+
+// NewTestClient creates a new test Helm client.
+func NewTestClient(conf ClientConfig) (Client, error) {
+	return &testClient{
+		releases:     make(map[string]storage.Release),
+		ClientConfig: conf,
+	}, nil
+}
+
+// Install installs a Helm chart and returns release information.
+func (c *testClient) Install(p InstallParameters) (storage.Release, error) {
+	c.releases[p.Name] = newRelease(p.Name, 0)
+	return c.releases[p.Name], nil
+}
+
+// List returns list of releases matching provided parameters.
+func (c *testClient) List(p ListParameters) ([]storage.Release, error) {
+	var releases []storage.Release
+	for _, v := range c.releases {
+		releases = append(releases, v)
+	}
+	return releases, nil
+}
+
+// Get returns a single release with the specified name.
+func (c *testClient) Get(name string) (storage.Release, error) {
+	release, ok := c.releases[name]
+	if !ok {
+		return nil, trace.NotFound("release %v not found", name)
+	}
+	return release, nil
+}
+
+// Upgrade upgrades a release.
+func (c *testClient) Upgrade(p UpgradeParameters) (storage.Release, error) {
+	release, ok := c.releases[p.Release]
+	if !ok {
+		return nil, trace.NotFound("release %v not found", p.Release)
+	}
+	c.releases[p.Release] = newRelease(p.Release, release.GetRevision()+1)
+	return c.releases[p.Release], nil
+}
+
+// Rollback rolls back a release to the specified version.
+func (c *testClient) Rollback(p RollbackParameters) (storage.Release, error) {
+	_, ok := c.releases[p.Release]
+	if !ok {
+		return nil, trace.NotFound("release %v not found", p.Release)
+	}
+	c.releases[p.Release] = newRelease(p.Release, p.Revision)
+	return c.releases[p.Release], nil
+}
+
+// Revisions returns revision history for a release with the provided name.
+func (c *testClient) Revisions(name string) ([]storage.Release, error) {
+	return nil, trace.NotImplemented("not implemented")
+}
+
+// Uninstall uninstalls a release with the provided name.
+func (c *testClient) Uninstall(name string) (storage.Release, error) {
+	release, ok := c.releases[name]
+	if !ok {
+		return nil, trace.NotFound("release %v not found", name)
+	}
+	delete(c.releases, name)
+	return release, nil
+}
+
+// Closer allows to cleanup the client.
+func (c *testClient) Close() error {
+	c.closed = true
```

Comment:
Not sure what this is so far useful for but I'd assert this in methods to catch early release or lack thereof?